src/lib/phdi-building-blocks/phdi_building_blocks/linkage.py

```diff
@@ -1,68 +1,100 @@
     """
-    Given a FHIR resource bundle (defined as a dictionary of resources
-    containing at least one patient resource):
-      - identify the patient resource(s) in the bundle
+    Given a FHIR resource bundle:
+      - identify all patient resource(s) in the bundle
       - extract standardized name, DOB, and address information for each
       - compute a unique hash string based on these fields
       - add the hash string to the list of identifiers held in that
         patient resource
-    This function assumes data has been standardized already by the
-    silver transforms.
+    :param dict bundle: The FHIR bundle for whose patients to add a
+    linking identifier
+    :param str salt_str: The suffix string added to prevent being
+    able to reverse the hash into PII
+    :param bool overwrite: Whether to write the new standardizations
+    directly into the given bundle, changing the original data (True
+    is yes)
+    :return: The bundle of data with patients having unique identifiers
+    :rtype: dict
     """
+    # Copy the data if we don't want to overwrite the original
+    if not overwrite:
+        bundle = copy.deepcopy(bundle)
 
-    for resource in bundle["entry"]:
-        if resource["resource"]["resourceType"] == "Patient":
-            patient = resource["resource"]
+    for resource in find_patient_resources(bundle):
+        patient = resource.get("resource")
 
-            # Combine given and family name
-            recent_name = next(
-                (name for name in patient["name"] if name.get("use") == "official"),
-                patient["name"][0],
-            )
+        # Combine given and family name
+        recent_name = get_field_with_use(patient, "name", "official", 0)
+        name_parts = recent_name.get("given", []) + [recent_name.get("family")]
+        name_str = "-".join([n for n in name_parts if n])
 
-            name_parts = recent_name.get("given", []) + [recent_name.get("family")]
-            name_str = "-".join([n for n in name_parts if n])
+        # Compile one-line address string
+        address_line = ""
+        if "address" in patient:
+            address = get_field_with_use(patient, "address", "home", 0)
+            address_line = get_one_line_address(address)
 
-            # Compile one-line address string
-            address_line = ""
-            if "address" in patient:
-                address = next(
-                    (addr for addr in patient["address"] if addr.get("use") == "home"),
-                    patient["address"][0],
-                )
-                address_line = " ".join(address.get("line", []))
-                address_line += f" {address.get('city')}, {address.get('state')}"
-                if "postalCode" in address and address["postalCode"]:
-                    address_line += f" {address['postalCode']}"
+        # Generate and store unique hash code
+        link_str = name_str + "-" + patient["birthDate"] + "-" + address_line
```

Comment:
Probably out of scope for this PR, but should we be enforcing some minimum quality criteria?  For example:
* Last name with at least 2 characters
* Valid birthdate
* Address line

There is a serious possibility if birthdate and address are missing, we could assign the the same ID to different patients.

Another edge case to worry about - hospitals often assign trauma names to patients who come in to the ER.  This could mean we get many patients with the same name (e.g. John Doe) and DOB.  Also these names will change to their actual name later in the visit and could trigger mismatches.

If others think this is worth recording, I can write up tickets.

Reply:
This is probably outside the scope of this PR, but is something we should do. I've added a TODO, and a fast follow to this PR should be to turn all of the TODOs into tickets.